posts/inside-rust/2020-03-02-recent-future-pattern-matching-improvements.md

```diff
@@ -0,0 +1,234 @@
+
+Suppose also that `Expr` has a field `span` that you would also use. In ancient times, that is before Rust 1.0, this was possible, but today, it results in an error:
+
+```rust
+error[E0303]: pattern bindings are not allowed after an `@`
+ --> src/lib.rs:L:C
+  |
+L |         bar @ Some(Expr { span }) => {}
+  |                           ^^^^ not allowed after `@`
+```
+
+This was turned into an error in [#16053], mainly due to the difficulties of encoding borrow checking rules in a sound way in the old AST based borrow checker.
+
+Since then, we have [removed the old borrow checker][rip_ast_borrowck] in favor of one based on [MIR], which is a simpler, and more appropriate data structure for borrow checking. Specifically, in the case of a statement like `let ref x @ ref y = a;`, we would get roughly the same MIR as if we had used `let x = &a; let y = &a;`.
+
+So now that having bindings to the right of `@` is handled uniformly and correctly by the borrow checker (e.g., the compiler won't allow `ref x @ ref mut y`), we have decided to allow them under [`#![feature(bindings_after_at)]`][tracking_at], now available on nightly. With the feature gate enabled, you may for example write:
+
+```rust
+#![feature(bindings_after_at)]
+
+fn main() {
+    if let x @ Some(y) = Some(0) {
+        dbg!(x, y);
+    }
+}
+```
+
+Our hope is that with providing this feature, we remove one surprising corner of the language.
+
+## Combining by-move and by-`ref` bindings
+
+[tracking_move_ref]: https://github.com/rust-lang/rust/pull/68376
+
+For similar reasons as noted in the case of bindings after `@`, Rust does not currently allow you to combine normal by-move bindings with those that are by-`ref`. For example, should you write...:
+
+```rust
+fn main() {
+    let tup = ("foo".to_string(), 0);
+    let (x, ref y) = tup;
+}
+```
+
+... you would get an error:
+
+```rust
+error[E0009]: cannot bind by-move and by-ref in the same pattern
+ --> src/main.rs:3:10
+  |
+3 |     let (x, ref y) = tup;
+  |          ^  ----- by-ref pattern here
+  |          |
+  |          by-move pattern here
+```
+
+At the same time, however, the compiler is perfectly happy to allow...:
+
+```rust
+fn main() {
+    let tup = ("foo".to_string(), 0);
+    let x = tup.0;
+    let ref y = tup.1;
+}
+```
+
+... even though there is no semantic difference between these programs.
+
+Now that we have moved to the new borrow checker, as outlined in the previous section, we have relaxed this restriction on nightly as well, so under [`#![feature(move_ref_pattern)]`][tracking_move_ref], you may write:
+
+```rust
+#![feature(move_ref_pattern)]
+
+fn main() {
+    let tup = ("foo".to_string(), 0);
+    let (x, ref y) = tup;
+}
+```
+
+## How can I help?
```

Comment:
I wonder if linking here or something from the top is worthwhile, i.e., have a call to action immediately, and then get into the background information.

Reply:
I like the idea. Perhaps with something like https://github.com/rust-lang/blog.rust-lang.org/pull/529#discussion_r387346540 would do?